### PR TITLE
Add ElasticsearchChatMessageHistory

### PR DESCRIPTION
**Pull request by demjened**
**Description**

This PR adds the `ElasticsearchChatMessageHistory` implementation that stores chat message history in the configured [Elasticsearch](https://www.elastic.co/elasticsearch/) deployment.

```python
from langchain.memory.chat_message_histories import ElasticsearchChatMessageHistory

history = ElasticsearchChatMessageHistory(
    es_url="https://my-elasticsearch-deployment-url:9200", index="chat-history-index", session_id="123"
)

history.add_ai_message("This is me, the AI")
history.add_user_message("This is me, the human")
```

**Dependencies**
- [elasticsearch client](https://elasticsearch-py.readthedocs.io/) required

### CODE REVIEW

File: libs/langchain/langchain/memory/chat_message_histories/elasticsearch.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    Args:
+        client: Elasticsearch client.
+        index: name of the index to use.
+        session_id: arbitrary key that is used to store the messages
+            of a single chat session.
+    """
+
+    def __init__(
+        self,
+        client: Elasticsearch,
```
Comment (joemcelroy) [inline libs/langchain/langchain/memory/chat_message_histories/elasticsearch.py:27]: i suggest keeping the same API as the ElasticsearchStore for connection. Using this function, you should be able to support a variety of connection configurations for local and cloud

https://github.com/langchain-ai/langchain/blob/master/libs/langchain/langchain/vectorstores/elasticsearch.py#L513

Comment (joemcelroy) [reply libs/langchain/langchain/memory/chat_message_histories/elasticsearch.py:27]: Also keep in mind we should customise the user-agent to "langchain-py-ms/{version}" to provide us telemetry.

Comment (demjened) [reply libs/langchain/langchain/memory/chat_message_histories/elasticsearch.py]: Updated `ElasticsearchChatMessageHistory` to use the same API and added user agent header.

File: libs/langchain/langchain/memory/chat_message_histories/elasticsearch.py [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+            self.client.index(
+                index=self.index,
+                document={
+                    "session_id": self.session_id,
+                    "created_at": round(time() * 1000),
+                    "history": json.dumps(_message_to_dict(message)),
+                },
+                refresh=True,
+            )
+        except ApiError as err:
```
Comment (baskaryan) [inline libs/langchain/langchain/memory/chat_message_histories/elasticsearch.py:175]: what's value of catching these? would most users want to handle errors themselves (eg maybe they want to retry adding the message)

Comment (joemcelroy) [reply libs/langchain/langchain/memory/chat_message_histories/elasticsearch.py:175]: yeh we should provide a better error message here (could not add message to elasticsearch for example) and continue to raise the error so users can handle the errors themselves.

File: libs/langchain/langchain/memory/chat_message_histories/elasticsearch.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+            items = []
+
+        return messages_from_dict(items)
+
+    def add_message(self, message: BaseMessage) -> None:
+        """Add a message to the chat session in Elasticsearch"""
+        try:
+            from elasticsearch import ApiError
+
+            print("indexing message", message)
```
Comment (joemcelroy) [inline libs/langchain/langchain/memory/chat_message_histories/elasticsearch.py:165]: either remove or replace with a log debug message instead.

File: libs/langchain/langchain/memory/chat_message_histories/elasticsearch.py [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+        try:
+            from elasticsearch import ApiError
+
+            self.client.delete_by_query(
+                index=self.index,
+                query={"term": {"session_id": self.session_id}},
+                refresh=True,
+            )
+        except ApiError:
+            logger.error("Could not clear session memory in Elasticsearch")
```
Comment (joemcelroy) [inline libs/langchain/langchain/memory/chat_message_histories/elasticsearch.py:189]: continue to raise error

File: libs/langchain/langchain/memory/chat_message_histories/elasticsearch.py (hunk 3 above)
Comment (demjened) [reply libs/langchain/langchain/memory/chat_message_histories/elasticsearch.py]: Oops, this was left there by mistake. I removed it.

File: libs/langchain/langchain/memory/chat_message_histories/elasticsearch.py (hunk 2 above)
Comment (demjened) [reply libs/langchain/langchain/memory/chat_message_histories/elasticsearch.py:174]: Added more verbose log message and statement to raise the error.

File: libs/langchain/langchain/memory/chat_message_histories/elasticsearch.py [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+        """Retrieve the messages from Elasticsearch"""
+        try:
+            from elasticsearch import ApiError
+
+            result = self.client.search(
+                index=self.index,
+                query={"term": {"session_id": self.session_id}},
+                sort="created_at:asc",
+            )
+        except ApiError as err:
```
[the remark this answers is not in the document]
Comment (demjened) [reply libs/langchain/langchain/memory/chat_message_histories/elasticsearch.py:146]: Done.

**Review by joemcelroy — APPROVED** [review]

**Comment by baskaryan** [comment]
lgtm, thank you @demjened @joemcelroy!

**Pull request merged**